omc3/knob_extractor.py

```diff
@@ -0,0 +1,459 @@
+        epilog=USAGE_EXAMPLES,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        prog="Knob Extraction Tool."
+    )
+)
+def main(opt) -> Optional[KnobsDict]:
+    """ Main knob extracting function. """
+    if not opt.time and not opt.state:
+        raise ValueError("No functionality selected. Set either `time` or `state`.")
+
+    ldb = pytimber.LoggingDB(source="nxcals")
+    time = _parse_time(opt.time, opt.timedelta)
+
+    if opt.state:
+        # only print the state of the StateTracker - the MetaState!
+        # I still don't know what this does,
+        # because I only get back that the variable does not exist. (jdilly)
+        state = ldb.get("LhcStateTracker:State", time)  # do first to have output together
+        LOGGER.info("---- STATE ------------------------------------")
+        LOGGER.info(state)
+        return None
+    
+    knobs_dict = _parse_knobs_defintions(opt.knob_definitions)
+    knobs_extract = _extract(ldb, knobs_dict, opt.knobs, time)
+    _write_knobsfile(opt.output, knobs_extract, time)
+    return knobs_extract
+
+
+def _extract(ldb, knobs_dict: KnobsDict, knob_categories: Sequence[str], time: datetime) -> KnobsDict:
+    """
+    Main function to gather data from  the state-tracker.
+
+    Args:
+        ldb: The pytimber database.
+        knobs_dict (KnobsDict): A mapping of all knob-names to KnobEntries.
+        knob_categories (Sequence[str]): Knob Categories or Knob-Names to extract.
+        time (datetime): The time, when to extract.
+
+    Returns:
+        Dict[str, KnobsDict]: Contains all the extracted knobs, grouped by categories.
+        When extraction was not possible, the value attribute of the respective KnobEntry is still None
+
+    """
+    LOGGER.info("---- EXTRACTING KNOBS -------------------------")
+    LOGGER.info(f"extracting knobs for {time}")
+    knobs = {}
+
+    LOGGER.info("---- KNOBS ------------------------------------")
+    for category in knob_categories:
+        for knob in KNOB_CATEGORIES.get(category, [category]):
+            knobs[knob] = knobs_dict[knob]
+
+            LOGGER.info(f"Looking for {knob:<34s} ")
+            knobkey = f"LhcStateTracker:{knob}:target"
+            knobvalue = ldb.get(knobkey, time.timestamp())  # use timestamp to preserve timezone info
+            if knobkey not in knobvalue:
+                LOGGER.warning(f"No value for {knob} found!")
+                continue
+
+            LOGGER.debug(f"Some value for {knob} extracted.")
+            timestamps, values = knobvalue[knobkey]
+            if len(values) == 0:
+                LOGGER.warning(f"No value for {knob} found for given time!")
+                continue
+
+            value = values[-1]
+            if not math.isfinite(value):
+                LOGGER.warning(f"Value for {knob} is not a number or infinite!")
+                continue
+
+            LOGGER.info(f"Knob value for {knob} extracted: {value} (unscaled)")
```

Comment:
Redundant with line 284 (`LOGGER.debug(f"Some value for {knob} extracted.")`). I would rather have just this one then, since it's the same but with the value.